File: phase-1-dsa-foundations/module-2-core-data-structures/graphs/GraphOperations.java
package graphs;

import java.util.*;

/**
 * Demonstrating Graph Representation and Traversal.
 * Using Adjacency List (Map<Integer, List<Integer>>).
 */
public class GraphOperations {

    static class Graph {
        // Map Node -> Neighbors
        private Map<Integer, List<Integer>> adjList;

        Graph() {
            adjList = new HashMap<>();
        }

        void addEdge(int src, int dest, boolean bidirectional) {
            adjList.putIfAbsent(src, new ArrayList<>());
            adjList.putIfAbsent(dest, new ArrayList<>());
            adjList.get(src).add(dest);
            if (bidirectional) {
                adjList.get(dest).add(src);
            }
        }

        void dfs(int start) {
            System.out.print("DFS: ");
            Set<Integer> visited = new HashSet<>();
            dfsHelper(start, visited);
            System.out.println();
        }

        private void dfsHelper(int curr, Set<Integer> visited) {
            visited.add(curr);
            System.out.print(curr + " ");
            for (int neighbor : adjList.getOrDefault(curr, new ArrayList<>())) {
                if (!visited.contains(neighbor)) {
                    dfsHelper(neighbor, visited);
                }
            }
        }

        void bfs(int start) {
            System.out.print("BFS: ");
            Set<Integer> visited = new HashSet<>();
            Queue<Integer> queue = new LinkedList<>();

            visited.add(start);
            queue.offer(start);

            while (!queue.isEmpty()) {
                int curr = queue.poll();
                System.out.print(curr + " ");

                for (int neighbor : adjList.getOrDefault(curr, new ArrayList<>())) {
                    if (!visited.contains(neighbor)) {
                        visited.add(neighbor);
                        queue.offer(neighbor);
                    }
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Graph g = new Graph();
        // 0 - 1
        // | |
        // 2 - 3
        g.addEdge(0, 1, true);
        g.addEdge(0, 2, true);
        g.addEdge(1, 3, true);
        g.addEdge(2, 3, true);

        System.out.println("--- Graph Traversals (Start 0) ---");
        g.dfs(0);
        g.bfs(0);
    }
}

File: go/phase-1-dsa-foundations/module-2-core-data-structures/graphs/graph_operations.go
package main

import "fmt"

type Graph struct {
	// AdjList: Node -> Neighbors
	AdjList map[int][]int
}

func NewGraph() *Graph {
	return &Graph{
		AdjList: make(map[int][]int),
	}
}

func (g *Graph) AddEdge(src, dest int, bidirectional bool) {
	g.AdjList[src] = append(g.AdjList[src], dest)
	
	// Ensure dest exists in map even if no outgoing edges
	if _, exists := g.AdjList[dest]; !exists {
		g.AdjList[dest] = []int{}
	}
	
	if bidirectional {
		g.AdjList[dest] = append(g.AdjList[dest], src)
	}
}

func (g *Graph) DFS(start int) {
	fmt.Print("DFS: ")
	visited := make(map[int]bool)
	g.dfsHelper(start, visited)
	fmt.Println()
}

func (g *Graph) dfsHelper(curr int, visited map[int]bool) {
	visited[curr] = true
	fmt.Printf("%d ", curr)
	
	for _, neighbor := range g.AdjList[curr] {
		if !visited[neighbor] {
			g.dfsHelper(neighbor, visited)
		}
	}
}

func (g *Graph) BFS(start int) {
	fmt.Print("BFS: ")
	visited := make(map[int]bool)
	queue := []int{start}
	visited[start] = true
	
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:] // Dequeue
		
		fmt.Printf("%d ", curr)
		
		for _, neighbor := range g.AdjList[curr] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}
	fmt.Println()
}

func main() {
	g := NewGraph()
	// 0 - 1
	// |   |
	// 2 - 3
	g.AddEdge(0, 1, true)
	g.AddEdge(0, 2, true)
	g.AddEdge(1, 3, true)
	g.AddEdge(2, 3, true)
	
	fmt.Println("--- Graph Traversals (Start 0) ---")
	g.DFS(0)
	g.BFS(0)
}
